File: day5_javascript/Day5Part1.js
const R = require('ramda');
const fs = require('fs');

const input = fs.readFileSync('input.txt', 'utf8').split('\n').map(number => parseInt(number));

function solve(input) {
  let instructions = R.clone(input);
  let index = 0;
  let count = 0;

  while (index < input.length) {
    const offset = instructions[index];
    instructions = R.update(index, R.inc(offset), instructions);
    index = R.add(index, offset);
    count = R.inc(count);
  }

  return count;
}

const result = solve(input);
console.log(result);

File: day5/Day5Part1.hs
import qualified Data.Sequence as S

solve :: Int -> Int -> S.Seq Int -> Int
solve index count xs =
  if index >= S.length xs
  then count
  else solve
    ((+) index $ S.index xs index)
    ((+) count 1)
    (S.update index ((+) 1 $ S.index xs index) xs)

main :: IO()
main = do
  input <- lines <$> readFile "input.txt"
  let instructions = S.fromList $ map (read :: String -> Int) input
  print $ solve 0 0 instructions